### PR TITLE
Migration to .NET Core 3.1

### LINKED ISSUE #23: Hosting library doesn't relevant to .net core

### LINKED ISSUE #30: Enable SSL enforcement into the framework

### LINKED ISSUE #35: Add support for .Net Core 3.1 to the framework
The upgrade should consider the following improvements:
- Possibility of changes in Json.Net
- Revisiting hosting model for windows service which is offered by the platform
- The default implementation of https and enforcing (UseHsts, UseHttpsRedirection
- Possible extension for Windows Authentication etc with AddNegotiate
- Changes related to health check such as MapHelathCheck and Authorization as AddAuthorization

### LINKED ISSUE #46: Exceptions are swallowed by the interception framework
I had an issue where i could not communicate from the angular client to my c# controller. Everything appeared correct, however i could not hit the break point in the controller. A generic exception was thrown that i couldn't see had any relevance to my code.

I eventually discovered that i had a genuine exception but the logging framework was trying to serialize everything to the logs. The serialization was failing on one of the object properties which was causing the error. 

A temporary work around was to reduce the logging level from Verbose to Warning and i was able to debug properly my code. Obviously i need to resolve what the exception is that is being swallowed, to do that the logging framework needs to be more resilient.

### LINKED ISSUE #62: Add support for Open API - test this with the security

### LINKED ISSUE #63: Add default home to the API

### LINKED ISSUE #64: Adding feature flag by default to the Easify

### LINKED ISSUE #65: Add validation to the options pattern

### PR DESCRIPTION
**Pull request by moattarwork**
This is the migration to version 3.1 of .NET Core. The full release note is mentioned in the release tag https://github.com/icgam/Easify/releases/tag/V3.0.0

- close #46 
- close #35 
- close #64 
- close #30 
- close #65 
- close #63 
- close #62 
- close #23 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

It has been tested through unit and integration tests. Also, the manual test has been done for Sample API and Sample Window Service to make sure about the usage and operation model.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/Easify.AspNetCore.Bootstrap/ApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -27,59 +29,128 @@
...
+                return null;
+            }
         }
 
-        private static void LogResolvedEnvironment(IHostingEnvironment env, ILoggerFactory loggerFactory)
+        private static void LogResolvedEnvironment(IHostEnvironment env, ILoggerFactory loggerFactory)
         {
             var log = loggerFactory.CreateLogger("Startup");
-            log.LogInformation($"Application is started in '{env.EnvironmentName.ToUpper()}' environemnt ...");
+            log.LogInformation($"Application is started in '{env.EnvironmentName.ToUpper()}' environment ...");
```
Comment (LewisPercival) [inline src/Easify.AspNetCore.Bootstrap/ApplicationBuilderExtensions.cs:153]: 👍

File: src/Easify.Resilience.UnitTests/Easify.Resilience.UnitTests.csproj [hunk 2]
```diff
@@ -1,19 +1,29 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
     <PropertyGroup>
-        <TargetFramework>netcoreapp2.2</TargetFramework>
 
         <IsPackable>false</IsPackable>
+
```
Comment (LewisPercival) [inline src/Easify.Resilience.UnitTests/Easify.Resilience.UnitTests.csproj:6]: some formatting here

File: src/Easify.Extensions/DateTimeProvider.cs [hunk 3]
```diff
@@ -24,5 +24,10 @@ public DateTime Now()
         {
             return DateTime.Now;
         }
+
+        public DateTime ToDay()
```
Comment (LewisPercival) [inline src/Easify.Extensions/DateTimeProvider.cs:28]: Today?

File: src/Easify.Configurations/InvalidConfigurationException.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+// 
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.Linq;
+using FluentValidation.Results;
+
+namespace Easify.Configurations
+{
```
Comment (LewisPercival) [inline src/Easify.Configurations/InvalidConfigurationException.cs:22]: some formatting in this file

File: src/Easify.RestEase.UnitTests/Easify.RestEase.UnitTests.csproj [hunk 5]
```diff
@@ -1,16 +1,23 @@
 <Project Sdk="Microsoft.NET.Sdk">
```
Comment (LewisPercival) [inline src/Easify.RestEase.UnitTests/Easify.RestEase.UnitTests.csproj:1]: here too

File: src/Samples/Easify.Sample.WebAPI.IntegrationTests/AuthenticationControllerTests.cs [hunk 6]
```diff
@@ -32,57 +32,53 @@ public sealed class AuthenticationControllerTests
         [Fact]
         public async Task GivenAuthenticationWithOAuth2_WhenQueryingTheSecureResourceWithoutAuthorizationHeader_ThenGetInternalServerError()
         {
-            using (var fixture = TestServerFixture<StartupForNoAuth>.Create())
-            {
-                // Arrange
-                // Act
-                var response = await fixture.Client.GetAsync("api/Authentication/secured");
+            // Arrange
+            using var fixture = TestApplicationFactory<StartupForNoAuth>.Create();
```
Comment (LewisPercival) [inline src/Samples/Easify.Sample.WebAPI.IntegrationTests/AuthenticationControllerTests.cs:36]: this syntax is cool too👍

**Review by LewisPercival — APPROVED** [review]
Looks great Mo! Not a whole lot of improvements to add from me, some minor formatting here and there. Nice uses of some newer c# features too. Looking forward to c#9!

File: src/Easify.Extensions/IDateTimeProvider.cs [hunk 7]
```diff
@@ -18,9 +18,9 @@
 
 namespace Easify.Extensions
 {
-    // TODO: Should be moved to Core project. Also Add Today as well.
     public interface IDateTimeProvider
     {
         DateTime Now();
+        DateTime ToDay();
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Extensions/IDateTimeProvider.cs:24]: Not sure about the name of this, Today would be better

File: src/Easify.Logging.SeriLog.LogEntries/LogEntriesExtensions.cs [hunk 8]
```diff
@@ -23,15 +24,18 @@ public static class LogEntriesExtensions
         public const string DefaultLogMessageTemplate =
             "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{MachineName}] [{EnvironmentUserName}] [{ProcessId}] [{UserName}] [{CorrelationId}] [{ThreadId}] [{Level}] {Message}{NewLine}{Exception}";
 
-        public static IProvideTemplate UseLogEntries(this ILoggerConfiguration configurationServices, string token)
+        public static ISinkBuilderContext UseLogEntries(this ISinkBuilderContext sinkBuilderContext, string token, Action<IProvideTemplate> configure)
         {
-            return new FluentLogEntriesSinkBuilder(configurationServices, token);
+            var builder = new FluentLogEntriesSinkBuilder(sinkBuilderContext, token);
+            configure(builder);
+            return builder.BuildAndCloneContext(sinkBuilderContext);
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Logging.SeriLog.LogEntries/LogEntriesExtensions.cs:31]: needs a blank line before return

File: src/Easify.Logging.SeriLog.LogEntries/LogEntriesExtensions.cs [hunk 9]
```diff
@@ -23,15 +24,18 @@ public static class LogEntriesExtensions
...
         }
 
-        public static IBuildSink UseLogEntries(this ILoggerConfiguration configurationServices,
+        public static ISinkBuilderContext UseLogEntries(this ISinkBuilderContext sinkBuilderContext,
             IConfigurationSection config)
         {
-            return new ConfigBasedLogEntriesSinkBuilder(configurationServices, config);
+            var builder = new LogEntriesSinkBuilder(sinkBuilderContext, config);
+            return builder.BuildAndCloneContext(sinkBuilderContext);
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Logging.SeriLog.LogEntries/LogEntriesExtensions.cs:38]: needs a blank line before return

File: src/Easify.Logging.SeriLog.LogEntries/LogEntriesSinkBuilder.cs [hunk 10]
```diff
@@ -21,23 +21,23 @@
...
+            _sinkBuilderContext = sinkBuilderContext ??
+                                     throw new ArgumentNullException(nameof(sinkBuilderContext));
             _config = config ?? throw new ArgumentNullException(nameof(config));
         }
 
         public LoggerConfiguration Build()
         {
             var options = _config.Get<LogEntriesConfiguration>();
-            return _configurationServices.SinkConfiguration.Logentries(options.Token,
+            return _sinkBuilderContext.LoggerConfiguration.WriteTo.Logentries(options.Token,
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Logging.SeriLog.LogEntries/LogEntriesSinkBuilder.cs:40]: needs a blank line before return

File: src/Easify.Logging.SeriLog.Loggly/LogglyExtensions.cs [hunk 11]
```diff
@@ -21,19 +21,24 @@ namespace Easify.Logging.SeriLog.Loggly
...
         }
 
-        public static IBuildSink UseLoggly(this ILoggerConfiguration configurationServices,
+        public static ISinkBuilderContext UseLoggly(this ISinkBuilderContext sinkBuilderContext,
             IConfigurationSection config)
         {
-            return new ConfigBasedLogglySinkBuilder(configurationServices, config);
-        }
+            var builder = new LogglySinkBuilder(sinkBuilderContext, config);
+            return builder.BuildAndCloneContext(sinkBuilderContext);
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Logging.SeriLog.Loggly/LogglyExtensions.cs:41]: needs a blank line before return

File: src/Easify.Logging.SeriLog.Seq/SeqExtensions.cs [hunk 12]
```diff
@@ -14,20 +14,24 @@
...
 {
     public static class SeqExtensions
     {
-        public static ISetApiKey UseSeq(this ILoggerConfiguration configurationServices, string serverUrl)
+        public static ISinkBuilderContext UseSeq(this ISinkBuilderContext sinkBuilderContext, string serverUrl, Action<ISetApiKey> configure)
         {
-            return new FluentSeqSinkBuilder(configurationServices, serverUrl);
+            var builder = new FluentSeqSinkBuilder(sinkBuilderContext, serverUrl);
+            configure(builder);
+            return builder.BuildAndCloneContext(sinkBuilderContext);
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Logging.SeriLog.Seq/SeqExtensions.cs:28]: needs a blank line before return

File: src/Easify.Logging.SeriLog.Seq/SeqExtensions.cs [hunk 13]
```diff
@@ -14,20 +14,24 @@
...
         }
 
-        public static IBuildSink UseSeq(this ILoggerConfiguration configurationServices, IConfigurationSection config)
+        public static ISinkBuilderContext UseSeq(this ISinkBuilderContext sinkBuilderContext, IConfigurationSection config)
         {
-            return new ConfigBasedSeqSinkBuilder(configurationServices, config);
+            var builder = new SeqSinkBuilder(sinkBuilderContext, config);
+            return builder.BuildAndCloneContext(sinkBuilderContext);
```
Comment (Justin-FrenchIndustries) [inline src/Easify.Logging.SeriLog.Seq/SeqExtensions.cs:34]: needs a blank line before return

**Review by Justin-FrenchIndustries — APPROVED** [review]
Looks really good. Only got comments around formatting (space before return statements, I prefer that, but we need to be consistent) and the name of a method didn't seem quite right

**Pull request merged**